### PR TITLE
implement vectorized evaluation for `builtinCharSig`

### PR DESCRIPTION
**Pull request by ZhaoQian888**
PCP #12106 
### What problem does this PR solve? 

implement vectorized evaluation for `builtinCharSig`
### What is changed and how it works?
``` 
BenchmarkVectorizedBuiltinStringFunc/builtinCharSig-VecBuiltinFunc-8                4639            241188 ns/op           22954 B/op       4097 allocs/op
BenchmarkVectorizedBuiltinStringFunc/builtinCharSig-NonVecBuiltinFunc-8             4228            272857 ns/op           22937 B/op       4096 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_string_vec.go [hunk 1]
```diff
@@ -2162,11 +2164,67 @@ func (b *builtinBitLengthSig) vecEvalInt(input *chunk.Chunk, result *chunk.Colum
...
+		}
+		tempres := string(b.convertToBytes(bigints))
+		charsetlabe := strings.ToLower(bufstr.GetString(i))
+		if bufstr.IsNull(i) || charsetlabe == "ascii" || strings.HasPrefix(charsetlabe, "utf8") {
+			result.AppendString(tempres)
+		} else {
+			encoding, charsetName := charset.Lookup(charsetlabe)
+			if encoding == nil {
+				result.AppendNull()
+				continue
```
Comment (wshwsh12) [inline expression/builtin_string_vec.go]: Should we return the error instead of appending null ?

File: expression/builtin_string_vec.go [hunk 2]
```diff
@@ -2162,11 +2164,67 @@ func (b *builtinBitLengthSig) vecEvalInt(input *chunk.Chunk, result *chunk.Colum
...
+			}
+			oldStr := tempres
+			tempres, _, err := transform.String(encoding.NewDecoder(), tempres)
+			if err != nil {
+				logutil.BgLogger().Warn("change charset of string",
+					zap.String("string", oldStr),
+					zap.String("charset", charsetName),
+					zap.Error(err))
+				result.AppendNull()
+				continue
```
Comment (wshwsh12) [inline expression/builtin_string_vec.go]: ditto

File: expression/builtin_string_vec.go (hunk 1 above)
Comment (ZhaoQian888) [reply expression/builtin_string_vec.go]: Yes.After return err the calculate will be interrupted.I will solve this problem

File: expression/builtin_string_vec.go [hunk 3]
```diff
@@ -2162,11 +2164,65 @@ func (b *builtinBitLengthSig) vecEvalInt(input *chunk.Chunk, result *chunk.Colum
...
+			if err != nil {
+				logutil.BgLogger().Warn("change charset of string",
+					zap.String("string", oldStr),
+					zap.String("charset", charsetName),
+					zap.Error(err))
+				return err
+			}
+			result.AppendString(tempres)
+		}
+
```
Comment (wshwsh12) [inline expression/builtin_string_vec.go]: Useless line.

File: expression/bench_test.go [hunk 4]
```diff
@@ -265,6 +265,29 @@ func (g *defaultGener) gen() interface{} {
 	return nil
 }
 
+// charInt64Gener is used to generate int which is equal to char's ascii
+type charInt64Gener struct{}
+
+func (g *charInt64Gener) gen() interface{} {
+	rand := time.Now().Nanosecond()
+	rand = rand % 128
```
Comment (SunRunAway) [inline expression/bench_test.go]: We may not need to limit it under 128, see https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_char
> CHAR() arguments larger than 255 are converted into multiple result bytes

I think we can limit it under 1024

Comment (SunRunAway) [inline expression/bench_test.go]: We may not need to limit it under 128, see https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_char
> CHAR() arguments larger than 255 are converted into multiple result bytes

I think we can limit it under 1024

**Pull request merged**